util/util.go

```diff
@@ -535,6 +536,121 @@ func (o MachinesByCreationTimestamp) Less(i, j int) bool {
+// or updates where Cluster.Status.InfrastructureReady transitions from false to true
+// This implements a common requirement for Machine infrastructure providers
+func WatchOnClusterUnPausedAndInfrastructureReady(c controller.Controller, mapFunc handler.Mapper, logger logr.Logger) error {
+	return c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: mapFunc,
+		},
+		predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldCluster := e.ObjectOld.(*clusterv1.Cluster)
+				newCluster := e.ObjectNew.(*clusterv1.Cluster)
+				log := logger.WithValues("predicate", "updateEvent", "namespace", newCluster.Namespace, "cluster", newCluster.Name)
+
+				switch {
+				// never return true for a paused Cluster
+				case newCluster.Spec.Paused:
+					log.V(4).Info("Cluster is paused, will not attempt to map associated resource.")
+					return false
+				// return true if Cluster.Status.InfrastructureReady has changed from false to true
+				case !oldCluster.Status.InfrastructureReady && newCluster.Status.InfrastructureReady:
+					log.V(4).Info("Cluster InfrastructureReady became ready, will attempt to map associated resource.")
+					return true
+				// return true if Cluster.Spec.Paused has changed from true to false
+				case oldCluster.Spec.Paused && !newCluster.Spec.Paused:
+					log.V(4).Info("Cluster was unpaused, will attempt to map associated resource.")
+					return true
+				// otherwise, return false
+				default:
+					log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+					return false
+				}
+			},
+			CreateFunc: func(e event.CreateEvent) bool {
+				cluster := e.Object.(*clusterv1.Cluster)
+				log := logger.WithValues("predicateEvent", "create", "namespace", cluster.Namespace, "cluster", cluster.Name)
+
+				// Only need to trigger a reconcile if the Cluster.Spec.Paused is false and
+				// Cluster.Status.InfrastructureReady is true
+				if !cluster.Spec.Paused && cluster.Status.InfrastructureReady {
+					log.V(4).Info("Cluster is not paused and has infrastructure ready, will attempt to map associated resource.")
+					return true
+				}
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				log := logger.WithValues("predicateEvent", "delete", "namespace", e.Meta.GetNamespace(), "cluster", e.Meta.GetName())
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				log := logger.WithValues("predicateEvent", "generic", "namespace", e.Meta.GetNamespace(), "cluster", e.Meta.GetName())
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+		},
+	)
+}
+
```

Comment:
Can we unify all of these function under `WatchOnClusterPaused` and update its documentation?

Reply:
If we need to change the function signature, let's do that if it makes sense (for the logger), we'll mark this as a ⚠️ and explain it in the release notes

Reply:
I can definitely consolidate WatchOnClusterPaused and WatchOnClusterUnPaused.

I'm not sure about also consolidating WatchOnClusterUnPausedAndInfrastructureReady, since it has different semantics and only a subset of controllers need to trigger reconciliation when the cluster infrastructure becomes ready. It would also break reconciliation when a Cluster is created with Cluster.Spec.Paused=false and Cluster.Status.InfrastructureReady=false for controllers that only care about Paused.

Reply:
What about having `WatchOnCluster` to be a generic function that takes in different predicates? Then we can have `WatchOnClusterPaused` call `WatchOnCluster` with the ClusterPausedPredicateFunc, and for infrastructure controllers that need the extra bits we can pass in a ClusterInfrastructureReadyPredicate as well

Reply:
@vincepri I like that idea and I'll work up a refactor around a base `WatchOnCluster`, however see https://github.com/kubernetes-sigs/cluster-api/pull/2877#discussion_r405042924 for why we might not want to treat infrastructure ready as a separate predicate

Reply:
I'm fine having examples / guidance to users of these functions, I'd also make sure to avoid setting up different watches with different predicates if possible, and instead having one that can be customized, wdyt?

Reply:
Refactored on top of a common `WatchOnCluster`.

I tried pulling out predicate functions, but realized that since we are casting to `clusterv1.Cluster` we'd need to have the ability to also return an error, so I avoided that. Happy to pull those back out if there is a reason we'd still want to be able to apply those separately, though I'm not sure there is a lot of value in being able to stack them with other predicates, since predicate handling with multiple predicates uses logical AND, and the way these predicates only return true when values change on Update.

I've also updated the godoc and added examples to help clarify their usage.

Reply:
This seems fine for now, we can clean it up later if there is a need